### DISCUSSION TITLE
XInput Controller, Set Configuration Response failed, Implementation Error or problem in usbstack??

### DISCUSSION THREAD

**Discussion by fluffymadness** [hathach/tinyusb#844] [Q&A] [answered]
Hi. Right now I'm trying to implement a xinput controller with tinyusb on an rp2040.
I've already put some hours into reading and debugging with no solution so far :/

From what I've read about xinput and seen so far it is basically a combination of control transfer (get for input report data, set for led data and rumble) 
plus hid reports for input report data, despite not having a hid descriptor.
see: https://www.partsnotincluded.com/understanding-the-xbox-360-wired-controllers-usb-data/

In tinyusb, I think i should be able to just use CFG_TUD_HID 1, instead of, CFG_TUD_VENDOR 1.

Interestingly when dumping with Thesycon usb descriptor dumper and comparing
the working example that i already have with lufa https://github.com/fluffymadness/ATMega32U4-XINPUT-Fightstick
The tinyusb one is giving me some requests failed for the string descriptor table. The rest of the descriptor however is
identical, so at least that part should work correctly.
```

Information for device XINPUT CONTROLLER (VID=0x045E PID=0x028E):

------------------------------
Connection Information:
------------------------------
Device current bus speed: FullSpeed
Device supports USB 1.1 specification
Device supports USB 2.0 specification
Device address: 0x000B
Current configuration value: 0x00
Number of open pipes: 0


------------------------------
Device Descriptor:
------------------------------
0x12	bLength
0x01	bDescriptorType
0x0200	bcdUSB
0xFF	bDeviceClass      (Vendor specific)
0xFF	bDeviceSubClass   
0xFF	bDeviceProtocol   
0x40	bMaxPacketSize0   (64 bytes)
0x045E	idVendor
0x028E	idProduct
0x0114	bcdDevice
0x01	iManufacturer
0x02	iProduct     
0x03	iSerialNumber
0x01	bNumConfigurations


-------------------------
Configuration Descriptor:
-------------------------
0x09	bLength
0x02	bDescriptorType
0x0099	wTotalLength   (153 bytes)
0x04	bNumInterfaces
0x01	bConfigurationValue
0x00	iConfiguration
0xA0	bmAttributes   (Bus-powered Device, Remote-Wakeup)
0xFA	bMaxPower      (500 mA)

Interface Descriptor:
------------------------------
0x09	bLength
0x04	bDescriptorType
0x00	bInterfaceNumber
0x00	bAlternateSetting
0x02	bNumEndPoints
0xFF	bInterfaceClass      (Vendor specific)
0x5D	bInterfaceSubClass   
0x01	bInterfaceProtocol   
0x00	iInterface

Unknown Descriptor:
------------------------------
0x11	bLength
0x21	bDescriptorType
Hex dump: 
0x11 0x21 0x00 0x01 0x01 0x25 0x81 0x14 0x00 0x00 
0x00 0x00 0x13 0x01 0x08 0x00 0x00 

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x81	bEndpointAddress  (IN endpoint 1)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x04	bInterval         (4 frames)

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x01	bEndpointAddress  (OUT endpoint 1)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x08	bInterval         (8 frames)

Interface Descriptor:
------------------------------
0x09	bLength
0x04	bDescriptorType
0x01	bInterfaceNumber
0x00	bAlternateSetting
0x04	bNumEndPoints
0xFF	bInterfaceClass      (Vendor specific)
0x5D	bInterfaceSubClass   
0x03	bInterfaceProtocol   
0x00	iInterface

Unknown Descriptor:
------------------------------
0x1B	bLength
0x21	bDescriptorType
Hex dump: 
0x1B 0x21 0x00 0x01 0x01 0x01 0x82 0x40 0x01 0x02 
0x20 0x16 0x83 0x00 0x00 0x00 0x00 0x00 0x00 0x16 
0x03 0x00 0x00 0x00 0x00 0x00 0x00 

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x82	bEndpointAddress  (IN endpoint 2)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x02	bInterval         (2 frames)

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x02	bEndpointAddress  (OUT endpoint 2)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x04	bInterval         (4 frames)

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x83	bEndpointAddress  (IN endpoint 3)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x40	bInterval         (64 frames)

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x03	bEndpointAddress  (OUT endpoint 3)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x10	bInterval         (16 frames)

Interface Descriptor:
------------------------------
0x09	bLength
0x04	bDescriptorType
0x02	bInterfaceNumber
0x00	bAlternateSetting
0x01	bNumEndPoints
0xFF	bInterfaceClass      (Vendor specific)
0x5D	bInterfaceSubClass   
0x02	bInterfaceProtocol   
0x00	iInterface

Unknown Descriptor:
------------------------------
0x09	bLength
0x21	bDescriptorType
Hex dump: 
0x09 0x21 0x00 0x01 0x01 0x22 0x84 0x07 0x00 

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x84	bEndpointAddress  (IN endpoint 4)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x10	bInterval         (16 frames)

Interface Descriptor:
------------------------------
0x09	bLength
0x04	bDescriptorType
0x03	bInterfaceNumber
0x00	bAlternateSetting
0x00	bNumEndPoints
0xFF	bInterfaceClass      (Vendor specific)
0xFD	bInterfaceSubClass   
0x13	bInterfaceProtocol   
0x04	iInterface

Unknown Descriptor:
------------------------------
0x06	bLength
0x41	bDescriptorType
Hex dump: 
0x06 0x41 0x00 0x01 0x01 0x03 

Microsoft OS Descriptor is not available. Error code: 0x0000001F


--------------------------------
String Descriptor Table
--------------------------------
Index  LANGID  String
0x00   0x0000  
0x01   0x0000  Request failed with 0x0000001F
0x02   0x0000  Request failed with 0x0000001F
0x03   0x0000  Request failed with 0x0000001F
0x04   0x0000  Request failed with 0x0000001F

------------------------------

Connection path for device: 
USB xHCI Compliant Host Controller
Root Hub
Generic USB Hub
Generic USB Hub
XINPUT CONTROLLER (VID=0x045E PID=0x028E) Port: 3

Running on: Windows 10 or greater (Build Version 19042)

Brought to you by TDD v2.17.0, Feb 23 2021, 14:04:02
```

Doing further investigation with wireshark shows me however, that it already fails at the SetConfiguration Response stage where on the atmega32u4 however it is successful.
I wonder if the SetConfiguration Request from the host requests something that is unsupported by the tinyusb stack ??
Or maybe I missed something crucial in my implementation.

Wireshark:
```
GET DESCRIPTOR Request DEVICE
GET DESCRIPTOR Response DEVICE
GET DESCRIPTOR Request CONFIGURATION
GET DESCRIPTOR Response CONFIGURATION
GET DESCRIPTOR Request CONFIGURATION
GET DESCRIPTOR Response CONFIGURATION
GET STATUS Request
GET STATUS Response
SET CONFIGURATION Request:

Frame 9: 36 bytes on wire (288 bits), 36 bytes captured (288 bits) on interface wireshark_extcap1888, id 0
    Interface id: 0 (wireshark_extcap1888)
        Interface name: wireshark_extcap1888
    Encapsulation type: USB packets with USBPcap header (152)
    Arrival Time: May 21, 2021 13:30:19.226524000 W. Europe Daylight Time
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1621596619.226524000 seconds
    [Time delta from previous captured frame: 0.000087000 seconds]
    [Time delta from previous displayed frame: 0.000087000 seconds]
    [Time since reference or first frame: 0.053640000 seconds]
    Frame Number: 9
    Frame Length: 36 bytes (288 bits)
    Capture Length: 36 bytes (288 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: usb]
USB URB
    [Source: host]
    [Destination: 1.12.0]
    USBPcap pseudoheader length: 28
    IRP ID: 0xffff870749fce6e0
    IRP USBD_STATUS: USBD_STATUS_SUCCESS (0x00000000)
    URB Function: URB_FUNCTION_SELECT_CONFIGURATION (0x0000)
    IRP information: 0x00, Direction: FDO -> PDO
        0000 000. = Reserved: 0x00
        .... ...0 = Direction: FDO -> PDO (0x0)
    URB bus id: 1
    Device address: 12
    Endpoint: 0x00, Direction: OUT
        0... .... = Direction: OUT (0)
        .... 0000 = Endpoint number: 0
    URB transfer type: URB_CONTROL (0x02)
    Packet Data Length: 8
    [Response in: 10]
    Control transfer stage: Setup (0)
Setup Data
    bmRequestType: 0x00
        0... .... = Direction: Host-to-device
        .00. .... = Type: Standard (0x0)
        ...0 0000 = Recipient: Device (0x00)
    bRequest: SET CONFIGURATION (9)
    bConfigurationValue: 1
    wIndex: 0 (0x0000)
    wLength: 0


SET CONFIGURATION Response:

Frame 10: 28 bytes on wire (224 bits), 28 bytes captured (224 bits) on interface wireshark_extcap1888, id 0
    Interface id: 0 (wireshark_extcap1888)
        Interface name: wireshark_extcap1888
    Encapsulation type: USB packets with USBPcap header (152)
    Arrival Time: May 21, 2021 13:30:19.230720000 W. Europe Daylight Time
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1621596619.230720000 seconds
    [Time delta from previous captured frame: 0.004196000 seconds]
    [Time delta from previous displayed frame: 0.004196000 seconds]
    [Time since reference or first frame: 0.057836000 seconds]
    Frame Number: 10
    Frame Length: 28 bytes (224 bits)
    Capture Length: 28 bytes (224 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: usb]
USB URB
    [Source: 1.12.0]
    [Destination: host]
    USBPcap pseudoheader length: 28
    IRP ID: 0xffff870749fce6e0
    IRP USBD_STATUS: USBD_STATUS_SET_CONFIG_FAILED (0xc0002000)
    URB Function: URB_FUNCTION_SELECT_CONFIGURATION (0x0000)
    IRP information: 0x01, Direction: PDO -> FDO
        0000 000. = Reserved: 0x00
        .... ...1 = Direction: PDO -> FDO (0x1)
    URB bus id: 1
    Device address: 12
    Endpoint: 0x00, Direction: OUT
        0... .... = Direction: OUT (0)
        .... 0000 = Endpoint number: 0
    URB transfer type: URB_CONTROL (0x02)
    Packet Data Length: 0
    [Request in: 9]
    [Time from request: 0.004196000 seconds]
    Control transfer stage: Complete (3)


```

I'm basically out of ideas.
Any help to get this working would be greatly appreciated. 
Here is my non-working code that i have so far:
https://github.com/fluffymadness/tinyusb-xinput-test

I'm using https://github.com/Wiz-IO/wizio-pico
and platformio on visual studio code for dev
with the baremetal configuration (no arduino core)

**Comment by zlittell** [comment]
You should definitely just be able to use CFG_TUD_HID, when I did this on the teensy I just used the standard HID functionality of the USB library. I should be able to look at this after hours today.

I did document the process of creating an xbox look alike controller, if its of any help.
https://zlittell.com/fightstick/introduction-to-the-project/
https://github.com/zlittell/MSF-XINPUT

**Reply by fluffymadness** [reply]
Already read that one quite some time ago :) Your project and library helped me a lot when i started that stuff.

**Comment by hathach** [comment]
Try to get the debug log of tinyusb by setting `#define CFG_TUSB_DEBUG           2`. By default it will use the printf, and you must have it retarget to uart and not cdc. I will give the hint on what failed

**Reply by fluffymadness** [reply]
Will do. Down below is the output I got:

USBD Setup Received 80 06 00 01 00 00 40 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes
USBD Bus Reset 

USBD Setup Received 00 05 13 00 00 00 00 00 
  Set Address
USBD Xfer Complete on EP 80 with 0 bytes

USBD Setup Received 80 06 00 01 00 00 12 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 02 00 00 FF 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 64 bytes
USBD Xfer Complete on EP 80 with 64 bytes
  Queue EP 80 with 64 bytes
USBD Xfer Complete on EP 80 with 64 bytes
  Queue EP 80 with 25 bytes
USBD Xfer Complete on EP 80 with 25 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 03 03 09 04 FF 00 
  Get Descriptor String[3]
  Queue EP 80 with 8 bytes
USBD Xfer Complete on EP 80 with 8 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 03 00 00 FF 00 
  Get Descriptor String[0]
  Queue EP 80 with 4 bytes
USBD Xfer Complete on EP 80 with 4 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 02 03 09 04 FF 00 
  Get Descriptor String[2]
  Queue EP 80 with 36 bytes
USBD Xfer Complete on EP 80 with 36 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 06 00 00 0A 00 
  Get Descriptor Device Qualifier
  Stall EP0

USBD Setup Received 80 06 00 01 00 00 12 00 
  Get Descriptor Device
  Queue EP 80 with 18 bytes
USBD Xfer Complete on EP 80 with 18 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 02 00 00 09 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 9 bytes
USBD Xfer Complete on EP 80 with 9 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 06 00 02 00 00 99 00 
  Get Descriptor Configuration[0]
  Queue EP 80 with 64 bytes
USBD Xfer Complete on EP 80 with 64 bytes
  Queue EP 80 with 64 bytes
USBD Xfer Complete on EP 80 with 64 bytes
  Queue EP 80 with 25 bytes
USBD Xfer Complete on EP 80 with 25 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 80 00 00 00 00 00 02 00 
  Get Status
  Queue EP 80 with 2 bytes
USBD Xfer Complete on EP 80 with 2 bytes
  Queue EP 00 with zlp Status
USBD Xfer Complete on EP 00 with 0 bytes

USBD Setup Received 00 09 01 00 00 00 00 00 
  Set Configuration
process_set_config 852: assert failed
process_control_request 643: assert failed
  Stall EP0
```

**Comment by fluffymadness** [comment]
Just adding return true in usbd.c gets it to run. The descriptor shows completely inclusive string descriptor table in the usb descriptor dumper and gamepad is also shown under the game controlls app in windows 10

```
    // Failed if cannot find supported driver
 ->   return true;
    TU_ASSERT(drv_id < TOTAL_DRIVER_COUNT);
```

**Reply by hathach** [reply]
your descriptor contain too many Vendor, although you don't use it, you still need to enable the CFG_TUD_VENDOR = 3 or more that match the descriptor. Even your application does not make use of those endpoints, it still need to be enabled and responded with NAK to host requests.

**Reply by fluffymadness** [reply]
First thank you for taking the time to help me.
I didn't get the above descriptor to work no matter how much I increased the CFG_TUD_VENDOR number.
These are the errors i get:

```
  Set Configuration
usbd_open_edpt_pair 1069: assert failed
vendord_open 189: assert failed
process_set_config 852: assert failed
process_control_request 643: assert failed
  Stall EP0
```
I now fetched the descriptor from my retrobit saturn controller which contains a bare minimum xinput descriptor
with only two endpoints, so in theory it should work without cfg_tud_vendor then ?

In case I also tested the small descriptor it with CFG_TUD_VENDOR 1 and CFG_TUD_VENDOR  2
I sadly still have the same issues.
Do i need some specific code to send the NAK requests you told me about in your previous post?

With the descriptor down below I temporary added the
->   return true; in usbd.c for testing. It appeared fine in the usb controllers app then. So the descriptor should be correct.
```
Information for device XINPUT CONTROLLER (VID=0x045E PID=0x028E):

------------------------------
Connection Information:
------------------------------
Device current bus speed: FullSpeed
Device supports USB 1.1 specification
Device supports USB 2.0 specification
Device address: 0x0038
Current configuration value: 0x01
Number of open pipes: 2


------------------------------
Device Descriptor:
------------------------------
0x12	bLength
0x01	bDescriptorType
0x0200	bcdUSB
0xFF	bDeviceClass      (Vendor specific)
0xFF	bDeviceSubClass   
0xFF	bDeviceProtocol   
0x40	bMaxPacketSize0   (64 bytes)
0x045E	idVendor
0x028E	idProduct
0x0572	bcdDevice
0x01	iManufacturer   "GENERIC"
0x02	iProduct        "XINPUT CONTROLLER"
0x03	iSerialNumber   "1.0"
0x01	bNumConfigurations


-------------------------
Configuration Descriptor:
-------------------------
0x09	bLength
0x02	bDescriptorType
0x0030	wTotalLength   (48 bytes)
0x01	bNumInterfaces
0x01	bConfigurationValue
0x00	iConfiguration
0x80	bmAttributes   (Bus-powered Device)
0xFA	bMaxPower      (500 mA)

Interface Descriptor:
------------------------------
0x09	bLength
0x04	bDescriptorType
0x00	bInterfaceNumber
0x00	bAlternateSetting
0x02	bNumEndPoints
0xFF	bInterfaceClass      (Vendor specific)
0x5D	bInterfaceSubClass   
0x01	bInterfaceProtocol   
0x00	iInterface

Unknown Descriptor:
------------------------------
0x10	bLength
0x21	bDescriptorType
Hex dump: 
0x10 0x21 0x10 0x01 0x01 0x24 0x81 0x14 0x03 0x00 
0x03 0x13 0x02 0x00 0x03 0x00 

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x81	bEndpointAddress  (IN endpoint 1)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x04	bInterval         (4 frames)

Endpoint Descriptor:
------------------------------
0x07	bLength
0x05	bDescriptorType
0x02	bEndpointAddress  (OUT endpoint 2)
0x03	bmAttributes      (Transfer: Interrupt / Synch: None / Usage: Data)
0x0020	wMaxPacketSize    (1 x 32 bytes)
0x08	bInterval         (8 frames)

Microsoft OS Descriptor is not available. Error code: 0x0000001F


--------------------------------
String Descriptor Table
--------------------------------
Index  LANGID  String
0x00   0x0000  0x0409 
0x01   0x0409  "GENERIC"
0x02   0x0409  "XINPUT CONTROLLER"
0x03   0x0409  "1.0"

------------------------------

Connection path for device: 
USB xHCI Compliant Host Controller
Root Hub
Generic USB Hub
Generic USB Hub
XINPUT CONTROLLER (VID=0x045E PID=0x028E) Port: 3

Running on: Windows 10 or greater (Build Version 19042)

Brought to you by TDD v2.17.0, Feb 23 2021, 14:04:02
```

**Reply by hathach** [reply]
From your descriptor, there is **no HID interface and only VENDOR**. I don't see using this exact descriptor could work with hid. Also for the vendor interface what is the `Unknown Descriptor:` in your descriptor, the builtin vendor couldn't make sense of it and cause the parser to panic. Vendor is fully custom, there is no way tinyusb could understand these all. For this you need to write your own driver using  `usbd_app_driver_get_cb()` . Here is how rpi team add their raspberrypi/pico-sdk#197
 
------------------------------
0x10	bLength
0x21	bDescriptorType
Hex dump: 
0x10 0x21 0x10 0x01 0x01 0x24 0x81 0x14 0x03 0x00 
0x03 0x13 0x02 0x00 0x03 0x00

**Comment by fluffymadness** [comment, marked as answer]
@hathach thx for pointing me to the pico example, it was a good starting point.
Everything is working now.
The working example can be found on my repo
https://github.com/fluffymadness/tinyusb-xinput

**Reply by hathach** [reply]
Great works, I am glad you got it working. Currently vendor interface parser is not supported, if it become more of popular issues, I would figure out a better way than having application to write their own driver later on.

**Comment by zlittell** [comment]
Definitely awesome stuff. Apologize I couldn't have been of any more assistance than linking things you had already seen. Really thought I would have free time over the weekend to assist, used all my free time on work :\

Will be interesting as tinyusb host mode develops more. Would be fun to mess around with forwarding the security packets from interface 3 to the controller to facilitate console usage. Something similar to what Brandon did here, but with a single chip: https://github.com/brandonlw/USBSimulator

**Reply by fluffymadness** [reply]
No worries. But please keep me posted if you manage to get the forwarding to work, just out of curiosity's sake, since I don't
have a Xbox360 or any modern xbox hardware to test/use it on.
